README.md
# Hacklet [![Build Status](https://travis-ci.org/mcolyer/hacklet.png)](https://travis-ci.org/mcolyer/hacklet)

A library, written in Ruby, for controlling the [Modlet][modlet] (smart) outlet.

If you haven't heard of the Modlet before, it's an outlet cover which
allows you to convert any standard US outlet into a smart outlet. This
means that you can control whether each socket is on or off as well as
determine how much energy it's using with a sampling frequency of 10
seconds.

There are alot of other similar products but this is the first one that
I've seen that [costs $50][amazon] and includes control as well as
monitoring of both sockets independently.

Checkout the companion project [hacklet-remote] if you're interested in
controlling your modlet with [IFTTT].

## Why

So why write another client?

Unfortunately the client software included with the device isn't
available on all platforms (Linux) and it's pretty heavyweight for what
it does.

The goal of this project is provide all the same functionality of the
bundled client but in a lightweight manner without dependencies on
external services.

## Installation

### Mac

```shell
brew install libftdi
gem install hacklet
```

### Linux (Ubuntu/Debian)

```shell
apt-get install libftdi1
gem install hacklet
echo 'ATTRS{idVendor}=="0403", ATTRS{idProduct}=="8c81", SUBSYSTEMS=="usb", ACTION=="add", MODE="0660", GROUP="plugdev"' | sudo tee /etc/udev/rules.d/99-thinkeco.rules
```

### Windows

**Note**: These directions haven't been confirmed. If you can confirm
them or improve them, please create an [issue][issues] or send a pull request.

1. Install [7zip] in order to extract Zadiag.
1. Download Zadig for [XP][zadiag-xp] or [later][zadiag-later].
1. Extract Zadig and run it.
1. Click `Options` -> `List all Devices` to populate the device list.
1. Open the device dropdown and look for the "Thinkeco" device and select it.
1. Select the libusbK driver and press the `Replace Driver` button.
1. Download and unzip [libftdi][libftdi-win]
1. Run `gem install hacklet`

If something isn't clear here, take a look at [libFTDI under Windows][libftdi-win-post] as this was summarized from there.

## Getting Started

```shell
# Add the device to the network, keep a copy of the network ids (ie 0xDEED)
hacklet commission

# Get the samples from the top socket on the device registered as 0xDEED
hacklet read -n 0xDEED -s 0

# Turn the top socket off
hacklet off -n 0xDEED -s 0

# And back on again
hacklet on -n 0xDEED -s 0
```

## Contributing

All contributions are welcome (bug reports, bug fixes, documentation or
new features)! All discussion happens using [issues] so if you are
interested in contributing:

* Search to make sure an issue doesn't already exist.
* If it doesn't, create a new issue and describe your proposal.

If you're interested in following the status of the project, simply
"watch" the repository on Github and you'll receive notices about all of
the new issues.

If your curious about how the hardware works or the specifics of the
protocol checkout out the [developer wiki].

### Contribution Workflow

* Fork the repository
* Install dependencies `bundle install`
* Create a feature branch `git checkout -b short-descriptive-name`
* Run tests `bundle exec rake`
* Write your feature (and tests)
* Run tests `bundle exec rake`
* Create a pull request

[modlet]: http://themodlet.com
[amazon]: http://www.amazon.com/gp/product/B00AAT43OA/ref=as_li_qf_sp_asin_tl?ie=UTF8&camp=1789&creative=9325&creativeASIN=B00AAT43OA&linkCode=as2&tag=matcol-20
[issues]: https://github.com/mcolyer/hacklet/issues
[developer wiki]: https://github.com/mcolyer/hacklet/wiki
[hacklet-remote]: https://github.com/mcolyer/hacklet-remote/
[IFTTT]: http://ifttt.com
[7zip]: http://www.7-zip.org/
[libftdi-win]: http://code.google.com/p/picusb/downloads/detail?name=libftdi1-1.0_devkit_mingw32_17Feb2013.zip
[zadiag-xp]: http://sourceforge.net/projects/libwdi/files/zadig/zadig_xp_v2.0.1.160.7z/download
[zadiag-later]: http://sourceforge.net/projects/libwdi/files/zadig/zadig_v2.0.1.160.7z/download
[libftdi-win-post]: http://embedded-funk.blogspot.com/2013/03/libftdi-under-windows.html


CHANGELOG.md

0.6.0 / 2013-06-18
==================

 * Add Cross-platform support by switching to libftdi-ruby from serialport.
 * API Change: Hacklet::Dongle.open is now used to create a session
   instead of #open_session.
 * Refactor dongle communication functions into SerialConnection for
   clarity.

0.5.1 / 2013-06-02
==================

 * Bugfix: Properly sets the time on the device after commissioning.

0.5.0 / 2013-05-30
==================

 * Can commission new modlets, read data from them and control them
   through a simple utility program. This is the minimal required in
   order to be useful.


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'hacklet'
require 'timecop'


spec/hacklet_spec.rb
describe Hacklet do
  it 'should have a version number' do
    Hacklet::VERSION.should_not be_nil
  end
end


spec/hacklet/dongle_spec.rb
require 'spec_helper'

describe Hacklet::Dongle do
  it 'can open a new session' do
    serial_port = double("Context").as_null_object

    # Boot
    serial_port.should_receive(:write_data).with([0x02, 0x40, 0x04, 0x00, 0x44].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x84, 0x16, 0x01,
      0x00, 0x00, 0x87, 0x03, 0x00, 0x30, 0x00, 0x33, 0x83, 0x69, 0x9A, 0x0B,
      0x2F, 0x00, 0x00, 0x00, 0x58, 0x4F, 0x80, 0x0A, 0x1C, 0x81].pack('c*'))

    # Boot Confirmation
    serial_port.should_receive(:write_data).with([0x02, 0x40, 0x00, 0x00, 0x40].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x80, 0x01, 0x10, 0xd1].pack('c*'))

    serial_port.should_receive(:usb_close)
    Ftdi::Context.should_receive(:new).and_return(serial_port)
    Hacklet::Dongle.open(Logger.new('/dev/null')) do
      # noop
    end
  end

  it 'can find a new device' do
    serial_port = double("Context").as_null_object

    # Boot
    serial_port.should_receive(:write_data).with([0x02, 0x40, 0x04, 0x00, 0x44].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x84, 0x16, 0x01,
      0x00, 0x00, 0x87, 0x03, 0x00, 0x30, 0x00, 0x33, 0x83, 0x69, 0x9A, 0x0B,
      0x2F, 0x00, 0x00, 0x00, 0x58, 0x4F, 0x80, 0x0A, 0x1C, 0x81].pack('c*'))

    # Boot Confirmation
    serial_port.should_receive(:write_data).with([0x02, 0x40, 0x00, 0x00, 0x40].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x80, 0x01, 0x10, 0xd1].pack('c*'))

    # Unlock Network
    serial_port.should_receive(:write_data).with([0x02, 0xA2, 0x36, 0x04, 0xFC, 0xFF, 0x90, 0x01, 0x02].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0xA0, 0xF9, 0x01, 0x00, 0x58].pack('c*'))

    # Ignored packet #1
    serial_port.should_receive(:read_data).and_return([0x02, 0x99, 0xd1, 0x23].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x01, 0xcc, 0x1f, 0x10, 0x00,
      0x00, 0x58, 0x4f, 0x80, 0x00, 0x77, 0x2a, 0x8a, 0x33, 0xa7, 0xf4, 0xf6,
      0x80, 0xd0, 0x9c, 0x5d, 0x3c, 0x84, 0xf5, 0x2b, 0x43, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xcb].pack('c*'))

    # Ignored packet #2
    serial_port.should_receive(:read_data).and_return([0x02, 0x98, 0xc0, 0x09].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0xcc, 0x1f, 0x10, 0x00, 0x00,
      0x58, 0x4f, 0x80, 0x00, 0x05].pack('c*'))

    # Useful response
    serial_port.should_receive(:read_data).and_return([0x02, 0xa0, 0x13, 0x0b].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x66, 0xad, 0xcc, 0x1f, 0x10, 0x00,
      0x00, 0x58, 0x4f, 0x80, 0x8e, 0xa9].pack('c*'))

    # Lock Network
    serial_port.should_receive(:write_data).with([0x02, 0xA2, 0x36, 0x04, 0xFC, 0xFF, 0x00, 0x01, 0x92].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0xA0, 0xF9, 0x01, 0x00, 0x58].pack('c*'))

    # Update Time
    serial_port.should_receive(:write_data).and_return([0x02, 0x40, 0x22, 0x06, 0x66, 0xAD, 0xdb, 0xb4, 0xa8, 0x51, 0x0b].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x22, 0x01, 0x00, 0x63].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0xA2, 0x03, 0x66, 0xAD, 0x00, 0x2a].pack('c*'))

    serial_port.should_receive(:usb_close)

    Ftdi::Context.should_receive(:new).and_return(serial_port)
    Hacklet::Dongle.open(Logger.new('/dev/null')) do |dongle|
      Timecop.freeze(Time.at(0x51a8b4db)) do
        dongle.commission
      end
    end
  end

  it 'can request a sample' do
    serial_port = double("Context").as_null_object

    # Boot
    serial_port.should_receive(:write_data).with([0x02, 0x40, 0x04, 0x00, 0x44].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x84, 0x16, 0x01,
      0x00, 0x00, 0x87, 0x03, 0x00, 0x30, 0x00, 0x33, 0x83, 0x69, 0x9A, 0x0B,
      0x2F, 0x00, 0x00, 0x00, 0x58, 0x4F, 0x80, 0x0A, 0x1C, 0x81].pack('c*'))

    # Boot Confirmation
    serial_port.should_receive(:write_data).with([0x02, 0x40, 0x00, 0x00, 0x40].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x80, 0x01, 0x10, 0xd1].pack('c*'))

    # Lock Network
    serial_port.should_receive(:write_data).with([0x02, 0xA2, 0x36, 0x04, 0xFC, 0xFF, 0x00, 0x01, 0x92].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0xA0, 0xF9, 0x01, 0x00, 0x58].pack('c*'))

    # Selecting the network
    serial_port.should_receive(:write_data).with([0x02, 0x40, 0x03, 0x04, 0xA7, 0xB4, 0x05, 0x00, 0x51].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x03, 0x01, 0x00, 0x42].pack('c*'))

    # Requesting the sample
    serial_port.should_receive(:write_data).with([0x02, 0x40, 0x24, 0x06, 0xA7, 0xB4,
      0x00, 0x01, 0x0A, 0x00, 0x7a].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x24, 0x01, 0x00, 0x65].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0xA4, 0x12].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0xA7, 0xB4, 0x00, 0x01, 0x0A,
      0x00, 0x69, 0x8D, 0x44, 0x51, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x1d].pack('c*'))

    serial_port.should_receive(:usb_close)

    Ftdi::Context.should_receive(:new).and_return(serial_port)
    Hacklet::Dongle.open(Logger.new('/dev/null')) do |dongle|
      dongle.lock_network
      dongle.select_network(0xA7B4)
      dongle.request_samples(0xA7B4, 0x0001)
    end
  end

  it 'can enable a socket' do
    serial_port = double("Context").as_null_object

    # Boot
    serial_port.should_receive(:write_data).with([0x02, 0x40, 0x04, 0x00, 0x44].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x84, 0x16, 0x01,
      0x00, 0x00, 0x87, 0x03, 0x00, 0x30, 0x00, 0x33, 0x83, 0x69, 0x9A, 0x0B,
      0x2F, 0x00, 0x00, 0x00, 0x58, 0x4F, 0x80, 0x0A, 0x1C, 0x81].pack('c*'))

    # Boot Confirmation
    serial_port.should_receive(:write_data).with([0x02, 0x40, 0x00, 0x00, 0x40].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x80, 0x01, 0x10, 0xd1].pack('c*'))

    # Lock Network
    serial_port.should_receive(:write_data).with([0x02, 0xA2, 0x36, 0x04, 0xFC, 0xFF, 0x00, 0x01, 0x92].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0xA0, 0xF9, 0x01, 0x00, 0x58].pack('c*'))

    # Selecting the network
    serial_port.should_receive(:write_data).with([0x02, 0x40, 0x03, 0x04, 0xA7, 0xB4, 0x05, 0x00, 0x51].pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x03, 0x01, 0x00, 0x42].pack('c*'))

    # Switching
    request = [0xff]*56
    request[5] = 0xA5
    request = [0x02, 0x40, 0x23, 0x3B, 0xA7, 0xB4, 0x00] + request + [0x11]
    serial_port.should_receive(:write_data).with(request.pack('c*'))
    serial_port.should_receive(:read_data).and_return([0x02, 0x40, 0x23, 0x01, 0x00, 0x62].pack('c*'))

    serial_port.should_receive(:usb_close)

    Ftdi::Context.should_receive(:new).and_return(serial_port)
    Hacklet::Dongle.open(Logger.new('/dev/null')) do |dongle|
      dongle.lock_network
      dongle.select_network(0xA7B4)
      dongle.switch(0xA7B4, 0x0000, true)
    end
  end
end


lib/hacklet/dongle.rb
require 'logger'
require 'timeout'

module Hacklet
  class Dongle
    attr_reader :logger

    # Public: Initializes a dongle instance and yields it.
    #
    # logger - The Logger instance to log to, defaults to STDOUT.
    #
    # Returns nothing.
    def self.open(logger=Logger.new(STDOUT))
      serial = SerialConnection.new(logger)
      dongle = Dongle.new(serial, logger)
      begin
        dongle.send(:boot)
        dongle.send(:boot_confirm)
        yield dongle
      ensure
        serial.close
      end
    end

    # serial - Serial connection to use with the dongle.
    # logger - The Logger instance to log to.
    def initialize(serial, logger)
      @serial = serial
      @logger = logger
    end

    # Public: Listens for new devices on the network.
    #
    # This must be executed within an open session.
    #
    # Returns nothing.
    def commission
      response = nil
      begin
        unlock_network
        Timeout.timeout(30) do
          @logger.info("Listening for devices ...")
          loop do
            buffer = @serial.receive(4)
            buffer += @serial.receive(buffer.bytes.to_a[3]+1)
            if buffer.bytes.to_a[1] == 0xa0
              response = BroadcastResponse.read(buffer)
              @logger.info("Found device 0x%x on network 0x%x" % [response.device_id, response.network_id])
              break
            end
          end
        end
      rescue Timeout::Error
      ensure
        lock_network
      end

      update_time(response.network_id) if response
    end

    # Public: Selects the network.
    #
    # This must be executed within an open session. I'm guessing it selects the
    # network.
    #
    # network_id - 2 byte identified for the network.
    #
    # Returns nothing.
    def select_network(network_id)
      @serial.transmit(HandshakeRequest.new(:network_id => network_id))
      HandshakeResponse.read(@serial.receive(6))
    end

    # Public: Request stored samples.
    #
    # network_id - 2 byte identified for the network.
    # channel_id - 2 byte identified for the channel.
    #
    # TODO: This needs to return a more usable set of data.
    # Returns the SamplesResponse.
    def request_samples(network_id, channel_id)
      @logger.info("Requesting samples")
      @serial.transmit(SamplesRequest.new(:network_id => network_id, :channel_id => channel_id))
      AckResponse.read(@serial.receive(6))
      buffer = @serial.receive(4)
      remaining_bytes = buffer.bytes.to_a[3] + 1
      buffer += @serial.receive(remaining_bytes)
      response = SamplesResponse.read(buffer)

      response.converted_samples.each do |time, wattage|
        @logger.info("#{wattage}w at #{time}")
      end
      @logger.info("#{response.sample_count} returned, #{response.stored_sample_count} remaining")

      response
    end

    # Public: Used to controls whether a socket is on or off.
    #
    # network_id - 2 byte identified for the network.
    # channel_id - 1 byte identified for the channel.
    # enabled    - true enables the socket and false disables it.
    #
    # Returns the SwitchResponse.
    def switch(network_id, channel_id, state)
      request = ScheduleRequest.new(:network_id => network_id, :channel_id => channel_id)
      if state
        request.always_on!
        @logger.info("Turning on channel #{channel_id} on network 0x#{network_id.to_s(16)}")
      else
        request.always_off!
        @logger.info("Turning off channel #{channel_id} on network 0x#{network_id.to_s(16)}")
      end
      @serial.transmit(request)
      ScheduleResponse.read(@serial.receive(6))
    end

    # Public: Unlocks the network, to add a new device.
    #
    # Returns the BootConfirmResponse
    def unlock_network
      @logger.info("Unlocking network")
      @serial.transmit(UnlockRequest.new)
      LockResponse.read(@serial.receive(6))
      @logger.info("Unlocking complete")
    end

    # Public: Locks the network, prevents adding new devices.
    #
    # Returns the BootConfirmResponse
    def lock_network
      @logger.info("Locking network")
      @serial.transmit(LockRequest.new)
      LockResponse.read(@serial.receive(6))
      @logger.info("Locking complete")
    end

  private
    # Private: Initializes the dongle for communication
    #
    # Returns the BootResponse
    def boot
      @logger.info("Booting")
      @serial.transmit(BootRequest.new)
      BootResponse.read(@serial.receive(27))
    end

    # Private: Confirms that booting was successful?
    #
    # Not sure about this.
    #
    # Returns the BootConfirmResponse
    def boot_confirm
      @serial.transmit(BootConfirmRequest.new)
      BootConfirmResponse.read(@serial.receive(6))
      @logger.info("Booting complete")
    end

    # Private: Updates the time of a device.
    #
    # This must be executed within an open session. I'm guessing it selects the
    # network.
    #
    # network_id - 2 byte identified for the network.
    #
    # Returns nothing.
    def update_time(network_id)
      @serial.transmit(UpdateTimeRequest.new(:network_id => network_id))
      UpdateTimeAckResponse.read(@serial.receive(6))
      UpdateTimeResponse.read(@serial.receive(8))
    end
  end
end


lib/hacklet/command.rb
require 'slop'
require 'logger'

module Hacklet
  class Command
    def self.run(dongle, arguments)
      Slop.parse(arguments, :help => true) do
        command 'on', :banner => 'Turn on the specifed socket' do
          on :n, :network=, 'The network id (ex. 0x1234)', :required => true
          on :s, :socket=, 'The socket id (ex. 0)', :required => true
          on :d, :debug, 'Enables debug logging' do
            dongle.logger.level = Logger::DEBUG
          end

          run do |opts, args|
            network_id = opts[:network][2..-1].to_i(16)
            socket_id = opts[:socket].to_i

            dongle.lock_network
            dongle.select_network(network_id)
            dongle.switch(network_id, socket_id, true)
          end
        end

        command 'off', :banner => 'Turn off the specifed socket' do
          on :n, :network=, 'The network id (ex. 0x1234)', :required => true
          on :s, :socket=, 'The socket id (ex. 0)', :required => true
          on :d, :debug, 'Enables debug logging' do
            dongle.logger.level = Logger::DEBUG
          end

          run do |opts, args|
            network_id = opts[:network][2..-1].to_i(16)
            socket_id = opts[:socket].to_i

            dongle.lock_network
            dongle.select_network(network_id)
            dongle.switch(network_id, socket_id, false)
          end
        end

        command 'read', :banner => 'Read all available samples from the specified socket' do
          on :n, :network=, 'The network id (ex. 0x1234)', :required => true
          on :s, :socket=, 'The socket id (ex. 0)', :required => true
          on :d, :debug, 'Enables debug logging' do
            dongle.logger.level = Logger::DEBUG
          end

          run do |opts, args|
            network_id = opts[:network][2..-1].to_i(16)
            socket_id = opts[:socket].to_i

            dongle.lock_network
            dongle.select_network(network_id)
            dongle.request_samples(network_id, socket_id)
          end
        end

        command 'commission', :banner => 'Add a new device to the network' do
          on :d, :debug, 'Enables debug logging' do
            dongle.logger.level = Logger::DEBUG
          end

          run do |opts, args|
            dongle.commission
          end
        end

        run do
          puts help
        end
      end
    end
  end
end


lib/hacklet/serial_connection.rb
require 'ftdi'

module Hacklet
  class SerialConnection
    def initialize(logger, port='/dev/ttyUSB0')
      @logger = logger
      @connection = Ftdi::Context.new
      @connection.usb_open(0x0403, 0x8c81)
      @connection.set_bitmode(0x00, Ftdi::BitbangMode[:reset])
      @connection.baudrate = 115200
      @connection.flowctrl = Ftdi::SIO_DISABLE_FLOW_CTRL
      @connection.dtr = 1
      @connection.rts = 1
      @receive_buffer = ""
    end

    # Public: Closes the connection
    #
    # Returns nothing.
    def close
      @connection.usb_close
    end

    # Public: Transmits the packet to the dongle.
    #
    # command - The binary string to send.
    #
    # Returns the number of bytes written.
    def transmit(command)
      @logger.debug("TX: #{unpack(command.to_binary_s).inspect}")
      @connection.write_data(command.to_binary_s)
    end

    # Public: Waits and receives the specified number of packets from the
    # dongle.
    #
    # bytes - The number of bytes to read.
    #
    # Returns a binary string containing the response.
    def receive(bytes)
      response = ""
      loop do
        if bytes <= @receive_buffer.bytesize
          array = @receive_buffer.bytes.to_a
          response = array[0..(bytes - 1)].pack('c*')
          @receive_buffer = array[(bytes)..-1].pack('c*')
          break
        end

        chunk = @connection.read_data
        if chunk.bytesize > 0
          @receive_buffer += chunk
        else
          sleep(0.1)
        end
      end
      @logger.debug("RX: #{unpack(response).inspect}")

      response
    end

  private
    # Private: Prints a binary string a concise hexidecimal form for debugging
    #
    # message - The message to parse.
    #
    # Returns a string of hexidecimal representing equivalent to the message.
    def unpack(message)
      message.unpack('H2'*message.size)
    end
  end
end


lib/hacklet/messages.rb
require 'bindata'

module Hacklet
  class Message < BinData::Record
    def calculate_checksum
      checksum_fields = @field_names - [:header, :checksum]
      buffer = StringIO.new
      io = BinData::IO.new(buffer)
      checksum_fields.each do |field|
        send(field).do_write(io)
      end
      buffer.rewind

      buffer.read.bytes.inject(0) { |s,x| s^x }
    end
  end

  class BootResponse < Message
    endian :big

    uint8  :header, :check_value => lambda { value == 0x02 }
    uint16 :command, :check_value => lambda { value == 0x4084 }
    uint8  :payload_length, :check_value => lambda { value == 22 }

    # TODO: Determine what's in here
    string :data, :length => 12

    uint64 :device_id

    # TODO: Determine what's in here
    uint16 :data2

    uint8 :checksum, :check_value => lambda { calculate_checksum == checksum }
  end

  class BootConfirmResponse < Message
    endian :big

    uint8  :header, :check_value => lambda { value == 0x02 }
    uint16 :command, :check_value => lambda { value == 0x4080 }
    uint8  :payload_length, :check_value => lambda { value == 1 }

    uint8  :data, :check_value => lambda { value == 0x10 }

    uint8 :checksum, :check_value => lambda { calculate_checksum == checksum }
  end

  class BroadcastResponse < Message
    endian :big

    uint8  :header, :check_value => lambda { value == 0x02 }
    uint16 :command, :check_value => lambda { value == 0xA013 }
    uint8  :payload_length, :check_value => lambda { value == 11 }

    uint16 :network_id
    uint64 :device_id

    # TODO: Not sure why this is here.
    uint8  :data

    uint8  :checksum, :check_value => lambda { calculate_checksum == checksum }
  end

  class LockResponse < Message
    endian :big

    uint8  :header, :check_value => lambda { value == 0x02 }
    uint16 :command, :check_value => lambda { value == 0xA0F9 }
    uint8  :payload_length, :check_value => lambda { value == 1 }

    uint8 :data, :check_value => lambda { value == 0x00 }

    uint8  :checksum, :check_value => lambda { calculate_checksum == checksum }
  end

  class UpdateTimeAckResponse < Message
    endian :big

    uint8  :header, :check_value => lambda { value == 0x02 }
    uint16 :command, :check_value => lambda { value == 0x4022 }
    uint8  :payload_length, :check_value => lambda { value == 1 }

    uint8  :data, :check_value => lambda { value == 0x00 }

    uint8  :checksum, :check_value => lambda { calculate_checksum == checksum }
  end

  class UpdateTimeResponse < Message
    endian :big

    uint8  :header, :check_value => lambda { value == 0x02 }
    uint16 :command, :check_value => lambda { value == 0x40a2 }
    uint8  :payload_length, :check_value => lambda { value == 3 }

    uint16 :network_id
    uint8  :data, :check_value => lambda { value == 0x00 }

    uint8  :checksum, :check_value => lambda { calculate_checksum == checksum }
  end

  class HandshakeResponse < Message
    endian :big

    uint8  :header, :check_value => lambda { value == 0x02 }
    uint16 :command, :check_value => lambda { value == 0x4003 }
    uint8  :payload_length, :check_value => lambda { value == 1 }

    uint8  :data, :check_value => lambda { value == 0x00 }

    uint8  :checksum, :check_value => lambda { calculate_checksum == checksum }
  end

  class AckResponse < Message
    endian :big

    uint8  :header, :check_value => lambda { value == 0x02 }
    uint16 :command, :check_value => lambda { value == 0x4024 }
    uint8  :payload_length, :check_value => lambda { value == 1 }

    uint8  :data, :check_value => lambda { value == 0x00 }

    uint8  :checksum, :check_value => lambda { calculate_checksum == checksum }
  end

  class SamplesResponse < Message
    endian :big

    uint8    :header, :check_value => lambda { value == 0x02 }
    uint16   :command, :check_value => lambda { value == 0x40A4 }
    uint8    :payload_length

    uint16   :network_id
    uint16   :channel_id

    # TODO: Confirm this is actually the device id
    uint16   :data

    uint32le :time
    uint8    :sample_count
    uint24le :stored_sample_count
    array    :samples, :type => [:uint16le], :initial_length => :sample_count

    uint8 :checksum, :check_value => lambda { calculate_checksum == checksum }

    def converted_samples
      t = time - 10
      samples.map { |s| t += 10; [Time.at(t), (s/13.0).round] }
    end
  end

  class ScheduleResponse < Message
    endian :big

    uint8  :header, :check_value => lambda { value == 0x02 }
    uint16 :command, :check_value => lambda { value == 0x4023 }
    uint8  :payload_length, :check_value => lambda { value == 1 }

    uint8  :data, :check_value => lambda { value == 0x00 }

    uint8  :checksum, :check_value => lambda { calculate_checksum == checksum }
  end

  class BootRequest < Message
    endian :big

    uint8  :header, :initial_value => 0x02
    uint16 :command, :initial_value => 0x4004
    uint8  :payload_length

    uint8  :checksum, :value => :calculate_checksum
  end

  class BootConfirmRequest < Message
    endian :big

    uint8  :header, :initial_value => 0x02
    uint16 :command, :initial_value => 0x4000
    uint8  :payload_length

    uint8  :checksum, :value => :calculate_checksum
  end

  class UnlockRequest < Message
    endian :big

    uint8  :header, :initial_value => 0x02
    uint16 :command, :initial_value => 0xA236
    uint8  :payload_length, :initial_value => 4

    # TODO: What is this?
    uint32 :data, :initial_value => 0xFCFF9001

    uint8  :checksum, :value => :calculate_checksum
  end

  class LockRequest < Message
    endian :big

    uint8  :header, :initial_value => 0x02
    uint16 :command, :initial_value => 0xA236
    uint8  :payload_length, :initial_value => 4

    # TODO: What is this?
    uint32 :data, :initial_value => 0xFCFF0001

    uint8  :checksum, :value => :calculate_checksum
  end

  class UpdateTimeRequest < Message
    endian :big

    uint8    :header, :initial_value => 0x02
    uint16   :command, :initial_value => 0x4022
    uint8    :payload_length, :initial_value => 6

    uint16   :network_id
    uint32le :time, :initial_value => lambda { Time.now.to_i }

    uint8    :checksum, :value => :calculate_checksum
  end

  class HandshakeRequest < Message
    endian :big

    uint8  :header, :initial_value => 0x02
    uint16 :command, :initial_value => 0x4003
    uint8  :payload_length, :initial_value => 4

    uint16 :network_id
    # TODO: What is this?
    uint16 :data, :initial_value => 0x0500

    uint8  :checksum, :value => :calculate_checksum
  end

  class SamplesRequest < Message
    endian :big

    uint8  :header, :initial_value => 0x02
    uint16 :command, :initial_value => 0x4024
    uint8  :payload_length, :initial_value => 6

    uint16 :network_id
    uint16 :channel_id
    # TODO: What is this?
    uint16 :data, :initial_value => 0x0A00

    uint8  :checksum, :value => :calculate_checksum
  end

  class ScheduleRequest < Message
    endian :big

    uint8  :header, :initial_value => 0x02
    uint16 :command, :initial_value => 0x4023
    uint8  :payload_length, :initial_value => 59

    uint16 :network_id
    uint8  :channel_id
    array  :schedule, :type => [:uint8], :initial_length => 56

    uint8  :checksum, :value => :calculate_checksum

    def always_off!
      bitmap = [0x7f]*56
      bitmap[5] = 0x25
      schedule.assign(bitmap)
    end

    def always_on!
      bitmap = [0xff]*56
      bitmap[5] = 0xa5
      schedule.assign(bitmap)
    end
  end
end
